### ISSUE TITLE
Sometimes an empty line is appended when processing a text selection with Notepad++

### ISSUE THREAD

**Issue by pidgeon777** [ascii-boxes/boxes#57]
A demonstration follows below. At first, there is only one empty line after the selected text, but when I process the selection through `"$(boxes_p)" -i box -t 4 -d c-cmt2 "$(ascfile)" "$(tempfile)"`, then a further empty line is added:

![b23wiwn63t](https://user-images.githubusercontent.com/13739892/52647373-e15e8500-2ee4-11e9-9ce9-c67e8b568303.gif)

The sample text I'm trying to enclose with Boxes:

````
	Basedsad
	as as dsa dsa readdirsad stackavail asd 
	as das das
	
			dfsdf
			sadfsdf
			sdfsdf
			
			   asdasd
			   asdasdff
			   f

  asd
    as fas 
	fas 
as
````
I'm using boxes version 1.2.

**Comment by tsjensen** [comment]
Thanks for the report. We'll have to find out if it's an NppExec issue or a *boxes* issue. Therefore, can you please share binary copies of the temp files? By "temp files" I am referring to the files referenced as `$(ascfile)` and `$(tempfile)` in the macro.

If you leave out the last line of the NppExec macro (the one that deletes the temp files), you can find them on disk. Also please share the exact NppExec macro code, so that I can try to reproduce the situation. Thank you!

**Comment by pidgeon777** [comment]
NppExec script run in Notepad++:

````
cls

cmd.exe /c exit %RANDOM%

set tempfile = $(SYS.TEMP)\NppBoxes_$(EXITCODE).out.txt
set ascfile = $(SYS.TEMP)\NppBoxes_$(EXITCODE).in.txt

sel_saveto $(ascfile) :a

"$(boxes_p)" -i box -t 4 -d c-cmt2 "$(ascfile)" "$(tempfile)"

sel_loadfrom $(tempfile)
cmd.exe /c del /f /q "$(tempfile)" "$(ascfile)"
````

Generated files:

[NppBoxes_28763.in.txt](https://github.com/ascii-boxes/boxes/files/2861320/NppBoxes_28763.in.txt)

[NppBoxes_28763.out.txt](https://github.com/ascii-boxes/boxes/files/2861319/NppBoxes_28763.out.txt)

**Comment by tsjensen** [comment]
Ok, thank you. I think I understand now. It appears to be a case of "works as designed", but prove me wrong.

*boxes* always assumes it is working on complete lines of text. That's just how the program works, and I think that's reasonable. If a line is not properly terminated with a newline character (can easily happen for example by `echo -n foo | boxes`), then *boxes* assumes it's being fed a line of text.  
Consequently, *boxes* outputs lines of text, each with their proper line endings.

In contrast to that, NppExec does not work on lines, but on chunks of text which are currently selected - these chunks can start anywhere, even square in the middle of a line.  
This is a mismatch of concepts between NppExec and *boxes* that we just need to be aware of.

So when you select less than a complete line in Notepad++, then *boxes* must add a line terminator at the end of the last line. To Notepad++, this is new, and causes the extra line break.

**TL/DR:** In order to avoid this problem, make sure that you always select *entire lines* before pressing the hotkey. "Entire lines" meaning "including their newlines". You can do that by placing the cursor at position 0 of a line, then selecting lines by pressing Shift+CursorDown.

**Comment by tsjensen** [comment]
Then again, the more I think about it, maybe we could make it the default behavior to have a newline character at the very end of *boxes* output only if there also was one at the very end of the input. This problem would be solved, but I need to think about it some more in order to see if it might cause other problems. Also, I don't know off the cuff if the input processing abstractions let us know about this, but I'll check.

**Comment by pidgeon777** [comment]
I think that the final newline character checking you proposed could do the trick. Hopefully, you'll have a chance to investigate more on this, it would be nice in my opinion to enhance this behaviour.

In any case, thank you.

**Comment by tsjensen** [comment]
I agree. I'll leave this open as a reminder. 😉

**Comment by pidgeon777** [comment]
Great, thank you!

**Comment by tsjensen** [comment]
Fixed in *boxes* 1.3, which was just released!
